homeassistant/components/recorder/history/modern.py

```diff
@@ -255,38 +317,64 @@ def get_full_significant_states_with_session(
 
 
 def _state_changed_during_period_stmt(
-    start_time: datetime,
-    end_time: datetime | None,
-    metadata_id: int | None,
+    start_time_ts: float,
+    end_time_ts: float | None,
+    single_metadata_id: int,
     no_attributes: bool,
     descending: bool,
     limit: int | None,
-) -> StatementLambdaElement:
-    stmt = _lambda_stmt_and_join_attributes(no_attributes, include_last_changed=False)
-    start_time_ts = start_time.timestamp()
-    stmt += lambda q: q.filter(
-        (
-            (States.last_changed_ts == States.last_updated_ts)
-            | States.last_changed_ts.is_(None)
+    include_start_time_state: bool,
+    run_start_ts: float | None,
+) -> Select | CompoundSelect:
+    stmt = (
+        _stmt_and_join_attributes(no_attributes, False)
+        .filter(
+            (
+                (States.last_changed_ts == States.last_updated_ts)
+                | States.last_changed_ts.is_(None)
+            )
+            & (States.last_updated_ts > start_time_ts)
         )
-        & (States.last_updated_ts > start_time_ts)
+        .filter(States.metadata_id == single_metadata_id)
     )
-    if end_time:
-        end_time_ts = end_time.timestamp()
-        stmt += lambda q: q.filter(States.last_updated_ts < end_time_ts)
-    if metadata_id:
-        stmt += lambda q: q.filter(States.metadata_id == metadata_id)
+    if end_time_ts:
+        stmt = stmt.filter(States.last_updated_ts < end_time_ts)
     if not no_attributes:
-        stmt += lambda q: q.outerjoin(
+        stmt = stmt.outerjoin(
             StateAttributes, States.attributes_id == StateAttributes.attributes_id
         )
-    if descending:
-        stmt += lambda q: q.order_by(States.metadata_id, States.last_updated_ts.desc())
-    else:
-        stmt += lambda q: q.order_by(States.metadata_id, States.last_updated_ts)
     if limit:
-        stmt += lambda q: q.limit(limit)
-    return stmt
+        stmt = stmt.limit(limit)
+    if not include_start_time_state or not run_start_ts:
+        return stmt.order_by(
+            States.metadata_id,
+            States.last_updated_ts.desc() if descending else States.last_updated_ts,
+        )
+
+    union_subquery = union_all(
+        _select_from_subquery(
+            _get_single_entity_start_time_stmt(
+                start_time_ts,
+                single_metadata_id,
+                no_attributes,
+                False,
+            ).subquery(),
+            no_attributes,
+            False,
+        ),
+        _select_from_subquery(
+            stmt.order_by(States.metadata_id, States.last_updated_ts).subquery(),
+            no_attributes,
+            False,
+        ),
+    ).subquery()
+    stmt = _select_from_subquery(union_subquery, no_attributes, False)
+    if not descending:
+        return stmt
+    # If descending, we need to reverse the results
```

Comment:
I think we could do this a bit faster in python to avoid the double sort but that would introduce some risk so I reverted that change. best to leave it as is as its probably not worth optimizing

Reply:
Postgresql will return wrong results if I don't which is why the newly added test is failing

I had to unrevert the refactor